mutiny-core/src/nostr/nwc.rs

```diff
@@ -333,6 +381,101 @@ impl NostrWalletConnect {
+                                &node_manager.logger,
+                            ) {
+                                Some(info) => info.status != HTLCStatus::Failed, // remove failed payments from budget
+                                None => true, // if we can't find the payment, keep it to be safe
+                            }
+                        });
+
+                        // we've cleaned a lot of storage, we should persist
+                        self.profile.spending_conditions =
+                            SpendingConditions::Budget(budget.clone());
+                        needs_save = true;
+
+                        // try again with cleaned budget
+                        if budget.sum_payments() + sats > budget.budget {
+                            Some("Budget exceeded.")
+                        } else {
+                            None
+                        }
+                    } else {
+                        None
+                    };
+
+                    let content = match budget_err {
+                        Some(err) => {
+                            log_warn!(node_manager.logger, "Attempted to exceed budget: {err}");
+                            Response {
+                                result_type: Method::PayInvoice,
+                                error: Some(NIP47Error {
+                                    code: ErrorCode::QuotaExceeded,
+                                    message: err.to_string(),
+                                }),
+                                result: None,
+                            }
+                        }
+                        None => {
+                            // attempt to pay invoice
+                            match self
+                                .pay_nwc_invoice(node_manager, from_node, &invoice)
+                                .await
+                            {
+                                Ok(resp) => {
+                                    // add payment to budget
+                                    budget.add_payment(&invoice);
+                                    self.profile.spending_conditions =
+                                        SpendingConditions::Budget(budget);
+                                    needs_save = true;
+                                    resp
+                                }
+                                Err(e) => {
+                                    // if the payment timed out, we should add it to the budget
+                                    // in case it is inflight and succeeds later
+                                    if let MutinyError::PaymentTimeout = e {
+                                        budget.add_payment(&invoice);
+                                        self.profile.spending_conditions =
+                                            SpendingConditions::Budget(budget);
+                                        needs_save = true;
+                                    }
```

Comment:
I think saving needs to happen first before the payment is attempted. I'm sure that may screw up how persistance of NWC profiles happens today but someone that opens the app, pulls down a pending nwc, then leaves before the 30s timeout will not have this counted towards their budget.

Reply:
Good point, added